### ISSUE TITLE
Drab + Webpack + React, "require is not defined"

### ISSUE THREAD

**Issue by fusillicode** [grych/drab#28]
Hi @grych I'm sorry for bothering again but unfortunately I'm encountering another client-side issue😞

I think it is related to my Webpack setup but as it is a pretty barebones and common configuration I'm opening this issue to make sure that there is no big problem underlying it. I hope indeed that it is another of my silly mistake 😅

Here's the error that I get from the console: 

![screen shot 2017-07-24 at 13 51 24](https://user-images.githubusercontent.com/2248534/28522477-712f3616-7078-11e7-8f0f-0157801a8f3d.png)

My configuration is the following:
`Elixir 1.4.5`
`Phoenix 1.3.rc2`
`Drab 0.5.0`

The code of the whole application can be found [here](https://github.com/fusillicode/metro/tree/ec9516a9fd258a6c51aed0edbe4be8d9caa645ce)
 
Please let me know if I can help you in some way 😄

Thanks again for all your commitment! 🤗

**Comment by grych** [comment]
Sorry, I can't help you with this, as it is definitely related to Webpack (and I don't know it). Drab injects only Javascript, so there should be no `require` statement in any of the Drab's JS code.

**Comment by fusillicode** [comment]
This is actually what I wanted to hear, i.e. "no problem with Drab"! 🤗

Thank you very much for your incredibly quick reply! 🚄

Feel free to close the issue (I'll just leave it here as I see you added the `help wanted` label) 😉

**Comment by grych** [comment]
Oh yes, obviously, I completely forgot.

Anyway, so far I don't know how to use it with Webpack, if it does not provide `require` as brunch does. Is there any guide how to use webpack with Phoenix sockets?

**Comment by grych** [comment]
Anyway it looks like you already have `require` provided by your JS. I just started your repo and here we are (from the console):

```
> require("phoenix").Socket
< function Socket(endPoint) {
    var _this4 = this;
    var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    _classCallCheck(this, Socket);
    this.stateChangeCallbacks = …
```

Also, looks like everything work for me...

**Comment by fusillicode** [comment]
There's no need to be sorry @grych! I'm the one that is bothering you! 😅

Anyway that's strange 🤔 . Just for curiosity, have you tried the `webpack` branch?

**Comment by grych** [comment]
Correct branch changes everything.
And it looks like `webpack` does not set `require` globally indeed.

I can't help with this. For now I can only say that Drab works with `branch` only.

### Anyway I will leave this opened to have this in mind:
- [ ] find a better way to import `Socket` from `phoenix_html`
- [ ] test it with `webpack`

**Comment by fusillicode** [comment]
> Correct branch changes everything.

😉 👍 

> And it looks like webpack does not set require globally indeed.

Unfortunately you're totally right 😔

> I can't help with this. For now I can only say that Drab works with branch only.

You helped me quite a lot! Moreover I find it totally legit to focus Drab on the default Phoenix build system, i.e. `brunch` 😄

Anyway I've already found an ugly hack:

1. attach the `Socket` to the `window` inside `app.js`
2. pull it out inside `Drab.js`. 

I'll try to think about a more clean and proof solution but for now I think that I'll go with it as I'm just building a prototype to try a few stuff around 😉 

Thanks once again for all your patience, support and commitment! The Elixir community is invaluable thanks to people like you! 😄 🍻

**Comment by OvermindDL1** [comment]
> find a better way to import Socket from phoenix_html

I'd recommend letting the user supply their own javascript expression (in the config or so?).

I tried to build a monolithic app and hit the same issue.  `Phoenix` was defined on `window`, but there is no such thing as `require` in such a build.  With such a config option it could default to `"require('phoenix').Socket"` and with my monolithic build I'd set it to `"Phoenix.Socket"`.

**Comment by GBH** [comment]
I'm seeing exactly same error and I'm actually using Brunch.

**Comment by grych** [comment]
@Kabie, I did not find the other way.
`Drab.Client` works with the JS eex template, which reads some info from conn, like the controller name, and from config.

**Comment by grych** [comment]
Closed with #133
